pymc3/step_methods/hmc/sgmcmc.py

```diff
@@ -0,0 +1,262 @@
+        step_size : float, default 0.25
+            Size of steps to take, automatically scaled down by 1/n**(1/4).
+            If step size adaptation is switched off, the resulting step size
+            is used. If adaptation is enabled, it is used as initial guess.
+        model : pymc3.Model
+            The model
+        kwargs: passed to BaseHMC
+
+        """
+        super(StochasticGradientFisherScoring, self).__init__(vars, **kwargs) 
+        self.B = B
+        
+            
+    def _initialize_values(self):
+        # 1. Initialize q, I_t at t = 0
+        # Sample from prior, merge parameters then append avg_I, t to the end
+        self.q_size = sum(v.dsize for v in self.vars) # dsize = int(np.prod(v.shape))
+        # Init I_t
+        avg_I = T.shared(np.asarray(np.zeros(q_param_size, q_param_size)), dtype=floatX)
+        self.q = None #self.independent_variables
+         
+        # 2. Set gamma
+        self.gamma = (self.batch_size + self.total_size)/(self.total_size)
+
+    def astep(self, ps0):
+        """
+        Returns a sequence 
+        of updates to members in ps0
+        """
+        n = self.batch_size
+        k = 1./n
+        avg_I = ps0.avg_I
+        t = ps0.t
+        B = self.B
+        q = ps0.q # collect only model parameters here
+
+        if B is None:
+            B = self.gamma * avg_I * self.total_size
+  
+        # logp_x_i log Probability of the model
+        logpt = self.logp_elemwise
+        # dlogp_x_i jacobian log probability of the model
+        dlogpt = self.dlogp_elemwise
+
+        #Gradient for each elem in X_n
+        g = dlogpt(q_params)
+
+        # 5. Calculate mean dlogp
+        avg_g = (1./n)*(np.sum(g, axis=0))
+        dlogp_t = np.sum(g)
+
+        # 6. Calculate approximate Fisher Score
+        V = (1./(n-1))*np.sum((g - avg_g)*(g - avg_g).T)
+
+        # 7. Update moving average
+        I_t = (1. - 1./self.t)*avg_I + (1./self.t)*V
+        
+        # 8. Noise Term
+        # The noise term is sampled from a normal distribution
+        # of mean 0 and std_dev = sqrt(4B/step_size)
+        # In order to generate the noise term, a standard
+        # normal dist. is scaled with 2B_ch/sqrt(step_size)
+        # where B_ch is cholesky decomposition of B
+        # i.e. B = dot(B_ch, B_ch^T)
+        B_ch = tensor.slinalg.cholesky(B)
+        noise_term = tensor.dot((2.*B_ch)/tensor.sqrt(self.step_size), \
+                trng.normal(self.q_size, 1))
+        # 9.
+        # Inv. Fisher Cov. Matrix
+        cov_mat = gamma * N * I_t + (4./self.lr) * B
+        inv_cov_mat = nlinalg.matrix_inverse(inv_cov_mat)
+        # Noise Coefficient
+        noise_coeff = (dlogp_t + N*avg_g + noise_term)
+        
+        q_t = q + 2*inv_cov_mat*noise_coeff
+
+
+        ps = SGFSFit(q=q_t, avg_I=I_t, t=t+1)
+
+        return ps
+
+    @staticmethod
+    def competence(var):
+        if var.dtype in continuous_types:
+            return Competence.IDEAL
```

Comment:
Should I recommend it to `Compatible` ? What difference will it cause at runtime ?